### PR TITLE
[IVY-2863] Tags UI Composables

### PR DESCRIPTION
**Pull request by Vishwa-Raghavendra**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [x] The PR is submitted to the `main` branch.
- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read the [Architecture Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Architecture.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] 🎬 I've attached a **screen recoding** of the changes. 

> Tip: drag & drop the video to the PR description.

## What's changed?

Describe with a few bullets **what's new:**
- a This PR adds the fix to display the tags on the homescreen tab
- b This PR adds the ability to filter transactions based on tags in ReportsScreen
- 

> 💡 Tip: Please, attach screenshots and screen recordings. It helps a lot!

## Risk Factors

**What may go wrong if we merge your PR?**
- N/A

**In what cases your code won't work?**

- a Tags Won't be seen on the main Screen

## Does this PR closes any GitHub Issues?

Check **[Ivy Wallet Issues](https://github.com/Ivy-Apps/ivy-wallet/issues)**.

- Closes #ISSUE_NUMBER

> Replace `ISSUE_NUMBER` with the number of the GitHub issue that this PR is fixing. If you've done that correctly, you'll see the issue title linked when previewing your PR description.

> **!Note:** Do **not** link the PR number. Link the number/id of the GitHub Issue from [issues](https://github.com/Ivy-Apps/ivy-wallet/issues).

## Troubleshooting CI failures ❌

GitHub Actions failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/data/src/main/java/com/ivy/data/repository/impl/TransactionRepositoryImpl.kt [hunk 1]
```diff
@@ -861,4 +877,10 @@ class TransactionRepositoryImpl @Inject constructor(
         accountId ?: return null
         return accountRepository.findById(AccountId(accountId))?.asset
     }
+
+    private suspend fun getTagsForTransactionIds(transactions: List<TransactionEntity>): Map<UUID, ImmutableList<Tag>> {
+        val transactionIdList = transactions.map { AssociationId(it.id) }
+        return tagReader.findByAssociatedId(transactionIdList).entries
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/impl/TransactionRepositoryImpl.kt:883]: We neee to memoize the tags reading in the `tagRepo` using the same pattern as in the `AccountRepository`

**Review by ILIYANGERMANOV — COMMENTED** [review]
1. Regarding changes in the legacy code idc.
2. Changes in `:shared:data` must not introduce new complexity and tech debt.

Left a few comments and have some suggestions:
1. A transaction should have `List<TagId>` not the real tag. Reason for that: if a tag is updated in db, you'll be forced to recompute all transactions and also duplication of the Tag instance leads to inconsistencies and also makes the Data layer slower.
2. Conversion to `ImmutableList` should be done in the UI layer; the data/domain layers should deal with domain models without caring about the UI needs

File: shared/base/src/main/java/com/ivy/base/legacy/LegacyTag.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.ivy.base.legacy
+
+import androidx.compose.runtime.Immutable
+import java.util.UUID
+
+@Immutable
+@Deprecated("Use Tag Data Model")
+@Suppress("DataClassTypedIDs")
+data class LegacyTag(val id: UUID, val name: String)
```
[the remark this answers is not in the document]
Comment (Vishwa-Raghavendra) [reply shared/base/src/main/java/com/ivy/base/legacy/LegacyTag.kt:9]: Leagacy Transaction Model and the new Transaction Model are in diff modules
as a result Tag data model is not accessible in the legacy module

[the remark this answers is not in the document]
Comment (ILIYANGERMANOV) [reply shared/base/src/main/java/com/ivy/base/legacy/LegacyTag.kt:9]: Can you make the legacy module depend on `share:data`? I would avoid polluting the new modules with legacy stuff if possible. Soon we'll have to delete all legacy ones and re-implement them properly using the new architecture. But if it's too much of a hassle, it's fine

File: shared/data/src/main/java/com/ivy/data/repository/impl/TagsRepositoryImpl.kt [hunk 3]
```diff
@@ -11,22 +11,31 @@ import com.ivy.data.repository.TagsRepository
...
-        return withContext(dispatchersProvider.io) {
+        return tagsMemo[id] ?: withContext(dispatchersProvider.io) {
             tagDao.findById(id.value)?.let {
                 with(mapper) { it.toDomain() }
-            }
+            }.also(::memoize)
         }
     }
 
+    override suspend fun findById(id: List<TagId>): List<Tag> {
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/impl/TagsRepositoryImpl.kt:35]: rename to `findByIds(ids: List<TagId>)`

File: shared/data/src/main/java/com/ivy/data/repository/impl/TagsRepositoryImpl.kt [hunk 4]
```diff
@@ -11,22 +11,31 @@ import com.ivy.data.repository.TagsRepository
...
+@Singleton
 class TagsRepositoryImpl @Inject constructor(
     private val mapper: TagMapper,
     private val tagDao: TagDao,
     private val writeTagDao: WriteTagDao,
     private val writeTagAssociationDao: WriteTagAssociationDao,
     private val dispatchersProvider: DispatchersProvider
 ) : TagsRepository {
+    private val tagsMemo = mutableMapOf<TagId, Tag>()
+    private var initialised: Boolean = false
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/impl/TagsRepositoryImpl.kt:25]: nit: rename to `findAllMemoized`

**Review by ILIYANGERMANOV — COMMENTED** [review]
The memoization looks good 💯

**Review by ILIYANGERMANOV — APPROVED** [review]

**Pull request merged**